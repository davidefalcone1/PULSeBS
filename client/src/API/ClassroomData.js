class ClassroomData{
    constructor(id, classroomName, maxSeats) {
        this.classId = id;
        this.classroomName = classroomName;
        this.maxSeats = maxSeats;
    }

    static fromJson(json){
        const temp =  Object.assign(new ClassroomData(), json);
        return temp;
    }
}

export default ClassroomData;