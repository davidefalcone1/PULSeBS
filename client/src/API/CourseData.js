//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;