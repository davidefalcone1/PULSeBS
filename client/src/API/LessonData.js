import moment from 'moment';

class LessonData {
    constructor(scheduleId, courseId, startDate, endDate, occupiedSeats, availableSeats,
        isLessonCancelled, isLessonRemote, classroom
        /*, normalBookings, cancelledBookings, waitingBookings */) {
        if (scheduleId)
            this.scheduleId = scheduleId;
        this.courseId = courseId;
        this.startDate = moment(new Date(startDate));
        this.endDate = moment(new Date(endDate));
        this.occupiedSeats = occupiedSeats;
        this.availableSeats = availableSeats;
        this.isLessonCancelled = isLessonCancelled;
        this.isLessonRemote = isLessonRemote;
        this.classroom = classroom

        //TODO use the correct variables
        this.normalBookings = 0;
        this.cancelledBookings = 0;
        this.waitingBookings = 0;
    }

    static fromJson(json) {
        const temp = Object.assign(new LessonData(), json);
        temp.startDate = moment(new Date(temp.startDate));
        temp.endDate = moment(new Date(temp.endDate));
        return temp;
    }
}

export default LessonData;