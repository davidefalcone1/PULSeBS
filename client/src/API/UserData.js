class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;