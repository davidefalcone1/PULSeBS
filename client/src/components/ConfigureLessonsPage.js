import React from 'react';
import ConfigureLessonsItem from './ConfigureLessonsItem';
import Button from 'react-bootstrap/esm/Button';
import Row from 'react-bootstrap/Row';
import Modal from 'react-bootstrap/Modal';
import Alert from 'react-bootstrap/Alert';
import Form from 'react-bootstrap/Form';
import ListGroup from 'react-bootstrap/ListGroup';
import moment from 'moment';
import { Redirect } from 'react-router-dom';
import { AuthContext } from '../_services/AuthContext';

const configureLessonsPage = (props) => {
  console.log(props);
  return(
    <ConfigureLessons lessonsList = {props.lessonsList} coursesList = {props.coursesList}
      classesList = {props.classesList} createNewLesson={props.createNewLesson}
      editLesson={props.editLesson} uploadFileLessons={props.uploadFileLessons}/>
  );
}

class ConfigureLessons extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.state = {
      isCreating: false, isEditing: false, file: undefined, isUploading: false, errorFile: false,
      scheduleId: '', courseId: 'Select course', lessonStatus: 'Select status', lessonType: 'Select type',
        startDate: '', endDate: '', classroom: 'Select classroom',
      errorCourseId: false, errorLessonStatus: false, errorLessonType: false,
        errorStartDate: false, errorEndDate: false, errorClassroom: false
    }
  }

  activateCreateModal = () => {
    this.setState({isCreating: true, 
      scheduleId: '', courseId: 'Select course', lessonStatus: 'Select status',
      lessonType: 'Select type', startDate: '', endDate: '', classroom: 'Select classroom',
    });
  }
  activateEditModal = (lesson) => {
    this.setState({isEditing: true,
      scheduleId: lesson.scheduleId, courseId: lesson.courseId, lessonStatus: lesson.isLessonCancelled,
      lessonType: lesson.isLessonRemote, startDate: lesson.startDate,
      endDate: lesson.endDate, classroom: lesson.classroom
    });
  }
  activateUploadFileModal = () => {
    this.setState({isUploading: true})
  }

  updateField = (name, value) => {
    this.setState({[name]: value}, () => {
      if(name === 'startDate' && moment(this.state.startDate).isAfter(moment(this.state.endDate))){
        this.setState({endDate: ''}, () => {})
      }
      if(this.state.courseId !== 'Select course'){
        this.setState({errorCourseId: false});
      }
      if(this.state.lessonStatus !== 'Select status'){
          this.setState({errorLessonStatus: false});
      }
      if(this.state.lessonType !== 'Select type'){
          this.setState({errorLessonType: false});
      }
      if(this.state.startDate !== ''){
          this.setState({errorStartDate: false});
      }
      if(this.state.endDate !== ''){
          this.setState({errorEndDate: false});
      }
      if(this.state.classroom !== 'Select classroom'){
          this.setState({errorClassroom: false});
      }
      if(this.state.file !== undefined && this.state.file !== ''){
        this.setState({errorFile: false});
      }
    });
  }

  handleSubmit = () => {
    if (!this.form.checkValidity()) {
        this.form.reportValidity();
    }
    else if(this.state.courseId === 'Select course'){
      this.setState({errorCourseId: true});
    }
    else if(this.state.lessonStatus === 'Select status'){
        this.setState({errorLessonStatus: true});
    }
    else if(this.state.lessonType === 'Select type'){
        this.setState({errorLessonType: true});
    }
    else if(this.state.startDate === ''){
        this.setState({errorStartDate: true});
    }
    else if(this.state.endDate === ''){
        this.setState({errorEndDate: true});
    }
    else if(this.state.classroom === 'Select classroom'){
        this.setState({errorClassroom: true});
    }
    else {
      if(this.state.isCreating)
        this.props.createNewLesson(this.state.courseId, this.state.errorLessonStatus,
          this.state.lessonType, this.state.startDate, this.state.endDate, this.state.classroom);
      else if(this.state.isEditing)
        this.props.editLesson(this.state.scheduleId, this.state.courseId, this.state.errorLessonStatus,
          this.state.lessonType, this.state.startDate, this.state.endDate, this.state.classroom);
    }
  }
  handleSubmitFile = () => {
    if (!this.formFile.checkValidity()) {
      this.formFile.reportValidity();
    }
    else if(this.state.file === undefined || this.state.file === ''){ 
      this.setState({errorFile: true});
    }
    else {
        this.props.uploadFileLessons(this.state.file)
    }
  }

  render(){
    return(
      <AuthContext.Consumer>
        {(context) => (
          <>        
            {context.user && this.props.coursesList && 
              <>
                <br/>
                <Row className="justify-content-around">
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateCreateModal();
                      }} id={"activateModalOfLessons"}>
                          Add New
                    </Button>
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateUploadFileModal();
                      }} id={"uploadFileOfLessons"}>
                          Upload File
                    </Button>
                </Row>
                <ListGroup as="ul" variant="flush">
                  <ListHeader />
                  {this.props.lessonsList.map((lesson) => 
                    <ConfigureLessonsItem key = {lesson.scheduleId} lesson = {lesson} 
                      courses = {this.props.coursesList} editLesson={this.activateEditModal}/>)
                  }
                </ListGroup>

                <Modal show={this.state.isCreating || this.state.isEditing} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>{this.state.isCreating ? 'Create new lesson' : 'Edit lesson'}</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newCourseForm" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmit();
                    }} ref={(form) => this.form = form}>
                      <Form.Group>
                        <Form.Label className="control-label">Course</Form.Label>
                        <Form.Control as="select" custom name="courseId" value = {this.state.courseId} 
                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                          <option>Select course</option>
                          {this.props.coursesList.map((course) =>
                            <option value={course.courseId}>{course.courseName}</option>
                          )}
                        </Form.Control>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Select Status</Form.Label>
                        <Form.Control as="select" custom name="lessonStatus" value = {this.state.lessonStatus} 
                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                          <option>Select status</option>
                          <option value={true}>Active</option>
                          <option value={false}>Cancelled</option>
                        </Form.Control>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Select Type</Form.Label>
                        <Form.Control as="select" custom name="lessonType" value = {this.state.lessonType} 
                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                          <option>Select type</option>
                          <option value={1}>In Presence</option>
                          <option value={0}>Remote</option>
                        </Form.Control>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Starting Date and Time</Form.Label>
                        <Form.Control type="datetime-local" name="startDate"
                          value = {this.state.startDate} min={moment().format("YYYY-MM-DD hh:mm")}
                          onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Ending Date and Time</Form.Label>
                        <Form.Control type="datetime-local" name="endDate"
                          value = {this.state.endDate} min={moment(this.state.startingDay).format("YYYY-MM-DD hh:mm")}
                          onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Classroom</Form.Label>
                        <Form.Control as="select" custom name="classroom" value = {this.state.classroom} 
                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                          <option>Select classroom</option>
                          {this.props.classesList.map((c) =>
                            <option value={c.classId}>{c.classroomName}</option>
                          )}
                        </Form.Control>
                      </Form.Group>
                      <Form.Group>
                        <div>
                          <button type="submit" className="btn btn-primary">
                            {this.state.isCreating ? 'Create' : 'Edit'}
                          </button>
                        </div>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isCreating: false, isEditing: false});
                      }}>Close</Button>
                      {this.state.errorCourseId &&
                        <>
                          <br/>
                          <Alert key="courseIdError" variant="danger">
                            Invalid course.
                          </Alert>
                        </>}
                      {this.state.errorLessonStatus &&
                        <>
                          <br/>
                          <Alert key="statusError" variant="danger">
                            Invalid status.
                          </Alert>
                        </>}
                      {this.state.errorLessonType &&
                        <>
                          <br/>
                          <Alert key="typeError" variant="danger">
                            Invalid type.
                          </Alert>
                        </>}
                      {this.state.errorStartDate &&
                        <>
                          <br/>
                          <Alert key="sdError" variant="danger">
                            Invalid starting date and time.
                          </Alert>
                        </>}
                      {this.state.errorEndDate &&
                        <>
                          <br/>
                          <Alert key="edError" variant="danger">
                            Invalid ending date and time.
                          </Alert>
                        </>}
                      {this.state.errorClassroom &&
                        <>
                          <br/>
                          <Alert key="crError" variant="danger">
                            Invalid classroom.
                          </Alert>
                        </>}
                  </Modal.Footer>
                </Modal>
              
                <Modal show={this.state.isUploading} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>Upload file</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newLessonFormFile" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmitFile();
                    }} ref={(form) => this.formFile = form}>                          
                      <Form.Group>
                        <Form.Label className="control-label">Insert file</Form.Label>
                        <Form.Control type="file" name="file" size = "lg"
                          required autoFocus accept=".csv"
                          onChange={(ev) => {                            
                            var f2 =function readFileContent(file) {
                              const reader = new FileReader()
                                return new Promise((resolve, reject) => {
                                  reader.onload=event=>resolve(event.target.result)
                                  reader.onerror = error => reject(error)
                                  reader.readAsText(file)
                                })
                            }
                            
                            f2(ev.target.files[0]).then(content => {
                              console.log(content)
                              this.updateField("file", content)
                            }).catch(error => console.log(error))
                          }}/>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isUploading: false});
                      }}>Close</Button>
                    {this.state.errorFile &&
                      <>
                        <br/>
                        <Alert key="fileError" variant="danger">
                          Invalid file.
                        </Alert>
                      </>}
                  </Modal.Footer>
                </Modal>
              </>
            }
            {context.user && !this.props.coursesList &&
              <NoItemsImage/>
            }
            {!context.user && <Redirect to="/login"/>}
          </>
        )}
      </AuthContext.Consumer>
    );
  };  
}


function ListHeader() {
  return(
    <ListGroup.Item id = {"lessonsList-header"}>
        <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
          <div className="col-sm-3">
              <h4>Course Name</h4>
          </div>
          <div className="col-sm-2">
              <h4>Start Date</h4>
          </div>
          <div className="col-sm-2">
              <h4>End Date</h4>
          </div>
          <div className="col-sm-2">
              <h4>Max Available Seats</h4>
          </div>
          <div className="col-sm-2">
              <h4>Classroom</h4>
          </div>
          <div className="col-sm-1">
              <h4>{' '}</h4>
          </div>
        </div>
    </ListGroup.Item>
  );
}
function NoItemsImage(props){
  return(
      <div className="col-sm-12 pt-3">
          <img width="800" height="600" className="img-button" src='./images/no_data_image.png' alt=""/>
      </div>
  );
}

export default configureLessonsPage;
