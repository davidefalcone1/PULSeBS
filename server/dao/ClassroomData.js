class ClassroomData{
    constructor(id, classroomName, maxSeats) {
        this.classId = id;
        this.classroomName = classroomName;
        this.maxSeats = maxSeats;
    }

}

module.exports = ClassroomData;