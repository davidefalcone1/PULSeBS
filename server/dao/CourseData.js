//usata per la retrieve della lista dei corsi
class CourseData{
    constructor(courseId, courseName, teacherId) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
    }
}

module.exports = CourseData;