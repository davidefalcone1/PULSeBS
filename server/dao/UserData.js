class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;