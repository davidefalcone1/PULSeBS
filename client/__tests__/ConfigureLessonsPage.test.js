import React from 'react';
import ConfigureLessonsPage from '../src/components/ConfigureLessonsPage';
import LessonData from '../src/API/LessonData';
import UserData from '../src/API/UserData';
import CourseData from '../src/API/CourseData';
import ClassroomData from '../src/API/ClassroomData';
import moment from 'moment';

Date.now = jest.fn(() => 1482363367071);//Avoid dynamic content on the page

describe('ConfigureLessonsPage', ()=>{
    let component;
    beforeAll(()=>{
        component = mount(
            <ConfigureLessonsPage classesList={[new ClassroomData(1, 'A1', 80)]} coursesList={[new CourseData('XY123', 'Software engineering 2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} lessonsList={[new LessonData(1, 'XY123', moment('2020-09-08T14:30'), moment('2020-09-08T17:00'), 9, 90, false, true, 'A1', 100, 4, 33, 0)]} createNewLesson={()=>{}} editLesson={()=>{}} uploadFileLessons={()=>{}}/>,
            {
                context: {user: new UserData(1, '12345', 'Davide Falcone', 'davidefalcone@polito.it', true)}
            }
        );
    });
    test('lesson modal', ()=>{
        component.find('#activateModalOfLessons').at(0).simulate('click');
        expect(component).toContainMatchingElement('#newCourseForm');
        const form = component.find('#newCourseForm').at(0);
        form.find('select').at(0).simulate('change', {target: {value: 'Software engineering 2'}});
        form.find('select').at(1).simulate('change', {target: {value: 'Active'}});
        form.find('select').at(2).simulate('change', {target: {value: 'Remote'}});
        form.find('select').at(3).simulate('change', {target: {value: '1 - 120'}});
        form.find('input').at(0).simulate('change', {target: {value: 'new-value'}});
        form.find('input').at(1).simulate('change', {target: {value: 'new-value'}});
        form.simulate('submit');
    });
    test('file lesson modal', ()=>{
        component.find('#uploadFileOfLessons').at(0).simulate('click');
        expect(component).toContainMatchingElement('#newLessonFormFile');
        const form = component.find('#newLessonFormFile').at(0);
        form.simulate('submit');
    });
});