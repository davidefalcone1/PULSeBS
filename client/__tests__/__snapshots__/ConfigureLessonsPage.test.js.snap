// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureLessonsPage Renders correctly 1`] = `
Array [
  <br />,
  <div
    className="justify-content-around row"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      id="activateModalOfLessons"
      onClick={[Function]}
      type="button"
    >
      Add New
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      id="uploadFileOfLessons"
      onClick={[Function]}
      type="button"
    >
      Upload File
    </button>
  </div>,
  <ul
    className="list-group list-group-flush"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="lessonsList-header"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-3"
        >
          <h4>
            Course Name
          </h4>
        </div>
        <div
          className="col-sm-2"
        >
          <h4>
            Start Date
          </h4>
        </div>
        <div
          className="col-sm-2"
        >
          <h4>
            End Date
          </h4>
        </div>
        <div
          className="col-sm-2"
        >
          <h4>
            Max Available Seats
          </h4>
        </div>
        <div
          className="col-sm-2"
        >
          <h4>
            Classroom
          </h4>
        </div>
        <div
          className="col-sm-1"
        >
          <h4>
             
          </h4>
        </div>
      </div>
    </div>
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="lesson-1"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-3"
        >
          <h6
            id="nameOfcourseForLesson_1"
          >
            Software engineering 2
          </h6>
        </div>
        <div
          className="col-sm-2"
        >
          <p
            id="startDateOfLesson1"
          >
            Tue 08-09-2020 14:30
          </p>
        </div>
        <div
          className="col-sm-2"
        >
          <p
            id="endDateOfLesson1"
          >
            Tue 08-09-2020 17:00
          </p>
        </div>
        <div
          className="col-sm-2"
        >
          <p
            id="seatsOfLesson1"
          >
            90
          </p>
        </div>
        <div
          className="col-sm-2"
        >
          <p
            id="classroomOfLesson1"
          >
            A1
          </p>
        </div>
        <div
          className="col-sm-1"
        >
          <button
            className="btn btn-primary"
            disabled={false}
            id="buttonFieldOflesson1"
            onClick={[Function]}
            type="button"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </ul>,
]
`;
